lib/services/matchers.go

```diff
@@ -57,6 +58,124 @@ func MatchResourceLabels(matchers []ResourceMatcher, resource types.ResourceWith
+	// the user is wanting to filter the contained resource ie. KubeClusters, Application, and Database.
+	switch filter.ResourceKind {
+	case types.KindNode:
+		specResource = resource
+
+	case types.KindKubeService:
+		return matchAndFilterKubeClusters(resource, filter)
+
+	case types.KindAppServer:
+		server, ok := resource.(types.AppServer)
+		if !ok {
+			return false, trace.BadParameter("expected types.AppServer, got %T", resource)
+		}
+		specResource = server.GetApp()
+
+	case types.KindDatabaseServer:
+		server, ok := resource.(types.DatabaseServer)
+		if !ok {
+			return false, trace.BadParameter("expected types.DatabaseServer, got %T", resource)
+		}
+		specResource = server.GetDatabase()
+
+	default:
+		return false, trace.NotImplemented("filtering for resource kind %q not supported", filter.ResourceKind)
+	}
+
+	return matchResourceByFilters(specResource, filter)
+}
+
+func matchResourceByFilters(resource types.ResourceWithLabels, filter MatchResourceFilter) (bool, error) {
+	if len(filter.PredicateExpression) > 0 {
+		parser, err := NewResourceParser(resource)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+
+		switch match, err := parser.EvalBoolPredicate(filter.PredicateExpression); {
+		case err != nil:
+			return false, trace.Wrap(err)
+		case !match:
+			return false, nil
+		}
+	}
+
+	if !types.MatchLabels(resource, filter.Labels) {
+		return false, nil
+	}
+
+	if !resource.MatchSearch(filter.SearchKeywords) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// matchAndFilterKubeClusters is similar to MatchResourceByFilters, but does two things in addition:
+//  1) handles kube service having a 1-N relationship (service-clusters)
+//     so each kube cluster goes through the filters
+//  2) filters out the non-matched clusters on the kube service and the kube service is
+//     updated in place with only the matched clusters
+//  3) only returns true if the service contained any matched cluster
+func matchAndFilterKubeClusters(resource types.ResourceWithLabels, filter MatchResourceFilter) (bool, error) {
+	server, ok := resource.(types.Server)
+	if !ok {
+		return false, trace.BadParameter("expected types.Server, got %T", resource)
+	}
+
+	kubeClusters := server.GetKubernetesClusters()
+
+	// Get the list of kube clusters.
+	resources := make([]types.ResourceWithLabels, len(kubeClusters))
+	for i, kube := range kubeClusters {
+		k8sV3, err := types.NewKubernetesClusterV3FromLegacyCluster(server.GetNamespace(), kube)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		resources[i] = k8sV3
+	}
+
+	// Apply filter to each kube cluster.
+	filtered := make([]*types.KubernetesCluster, 0, len(kubeClusters))
+	for i, resource := range resources {
+		switch match, err := matchResourceByFilters(resource, filter); {
+		case err != nil:
+			return false, trace.Wrap(err)
+		case match:
+			filtered = append(filtered, kubeClusters[i])
+		}
+	}
```

Comment:
I found the original code a bit confusing, this reads better IMO.

```suggestion
	// Apply filter to each kube cluster.
	filtered := make([]*types.KubernetesCluster, 0, len(kubeClusters))
	for _, kube := range kubeClusters {
		kubeResource, err := types.NewKubernetesClusterV3FromLegacyCluster(server.GetNamespace(), kube)
		if err != nil {
			return false, trace.Wrap(err)
		}
		match, err := matchResourceByFilters(kubeResource, filter)
		if err != nil {
			return false, trace.Wrap(err)
		}
		if match {
			filtered = append(filtered, kube)
		}
	}
```